File: chap07_calcium/L_type_Ca_3state.m

%  -------------------------------------------------------------------
%
%   3-state model of a L-type calcium channel
%
%   For Chapter 7, Section 7.4.2 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd.
%
%  -------------------------------------------------------------------

function L_type_Ca_3state

close all
clear all
clc
set(0,                           ...
   'defaultaxesfontsize', 18,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0);

%parameters
p.f=0.85;
p.omega=0.02;
p.g=2;

Vmlist=[-75,  -55, -35, -15, 5,25,45];
for j=1:length(Vmlist)
    Vm=Vmlist(j);

    p.alpha=2*exp(0.012*(Vm-35));
    p.beta = 0.0882*exp(-0.05*(Vm-35));
    c=0;
    p.gamma = 0.44*c;

    p.fbar = p.alpha*p.f/(p.alpha+p.beta);
    p.omegabar = p.omega*(p.beta/2+p.alpha)/(2*p.alpha+p.beta/2);

    dt = 0.01;
    tend=10;
    init = [0.0102,0.9898];
    tspan = [0:dt:tend];
    [t,sol] = ode15s(@(t,x)Ltrhs(t,x,p),tspan,init);
    op=sol(:,1);
    keep(:,j) = op;   % for external plotting

    figure(1)
    plot(t,op )
    hold on
    xlabel('Time (ms)')
    ylabel('Open probability')
end
legend('boxoff')
legend('V=-75','V=-55','V=-35','V=-15','V=5','V=25','V=45')
set(gca,'linewidth',1.5)
box off
hold off

end % of main

%writematrix([tspan', keep],'Ltype.dat')

%%
function out=Ltrhs(t,x,p)

c=(t>3)*2;
p.gamma = 0.44*c;
p.gammabar = p.gamma*(p.beta+2*p.alpha)/(p.alpha+p.beta);
op=x(1);
n3=x(2);
c3=1-n3-op;

opp = -p.g*op+p.fbar*n3;
n3p=-(p.fbar+p.gammabar)*n3 +p.omegabar*c3+p.g*op;

out=[opp;n3p];
end

File: chap07_calcium/L_type_Ca_3state.py
#   -------------------------------------------------------------------
# 
#    3-state model of a L-type calcium channel
# 
#    For Chapter 7, Section 7.4.2 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd.
# 
#   -------------------------------------------------------------------

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt

def Ltrhs(t, x):
    c = 2 if t > 3 else 0
    gamma = 0.44 * c
    gammabar = gamma * (beta + 2 * alpha) / (alpha + beta)
    
    op = x[0]
    n3 = x[1]
    c3 = 1 - n3 - op
    
    opp = -g * op + fbar * n3
    n3p = -(fbar + gammabar) * n3 + omegabar * c3 + g * op
    
    return [opp, n3p]


f = 0.85
omega = 0.02
g = 2

Vmlist = [-75, -55, -35, -15, 5, 25, 45]
keep = []

plt.figure()
for Vm in Vmlist:
    alpha = 2 * np.exp(0.012 * (Vm - 35))
    beta = 0.0882 * np.exp(-0.05 * (Vm - 35))
    c = 0
    gamma = 0.44 * c
    
    fbar = alpha * f / (alpha + beta)
    omegabar = omega * (beta / 2 + alpha) / (2 * alpha + beta / 2)
    
    dt = 0.01
    tend = 10
    init = [0.0102, 0.9898]
    tspan = np.arange(0, tend + dt, dt)
    
    sol = solve_ivp(lambda t, x: Ltrhs(t, x), [0, tend], init, t_eval=tspan, method='LSODA')
    op = sol.y[0, :]
    keep.append(op)
    
    plt.plot(sol.t, op, label=f'V={Vm}')


plt.xlabel('t (ms)')
plt.ylabel('Open probability,o')
plt.legend()


